### PR TITLE
15 edit and delete options

### PR DESCRIPTION
**Pull request by divino**
added fixes for issue #15

### CODE REVIEW

**Comment by divino** [comment]
@jenbreese - may I ask for review on the ui changes specially the changes on the details page?

**Comment by jenbreese** [comment]
@divino, sure I will take a look.

File: app/templates/bower.json [hunk 1]
```diff
@@ -23,6 +23,7 @@
     "vkbeautify-wrapper": "*",
     "highcharts": "^4.2",
     "angular-google-maps": "2.3.2",
+    "angular-messages": "^1.5.7",
```
Comment (grtjn) [inline app/templates/bower.json]: Please use "~1.4.4" instead

File: app/templates/bower.json [hunk 2]
```diff
@@ -37,6 +38,6 @@
   },
   "private": true,
   "resolutions": {
-    "angular": "~1.4.4"
+    "angular": "~1.5.7"
```
Comment (grtjn) [inline app/templates/bower.json]: please undo this change..

File: app/templates/rest-api/config/options/all.xml [hunk 3]
```diff
@@ -92,6 +92,15 @@
     </range>
   </constraint>
 
+  <constraint name="docFormat">
+    <range type="xs:string" facet="true" collation="http://marklogic.com/collation/codepoint">
+      <facet-option>limit=5</facet-option>
+      <facet-option>item-order</facet-option>
+      <facet-option>descending</facet-option>
+      <path-index>//docFormat</path-index>
```
Comment (grtjn) [inline app/templates/rest-api/config/options/all.xml]: You can leave off the // in path index expressions..

File: app/templates/ui/app/app.js [hunk 4]
```diff
@@ -17,7 +17,9 @@
     'ui.bootstrap',
     'ui.router',
     'ui.tinymce',
-    'ngToast'
+    'cb.x2js',
+    'ngToast',
+    'ngMessages'
```
Comment (grtjn) [inline app/templates/ui/app/app.js]: I think cb.x2js and ngMessages dependencies belong in create.module.js

File: app/templates/ui/app/create/create.controller.js [hunk 5]
```diff
@@ -44,20 +81,42 @@
...
-        toast.danger(response.data);
-      });
+      var extension = '.json';
+      var data = ctrl.person;
+      if (ctrl.person.docFormat == 'xml') {
+        extension = '.xml';
+        var wrap = {
+          xml : ctrl.person
+        };
+        data = x2js.json2xml_str(wrap);
```
Comment (grtjn) [inline app/templates/ui/app/create/create.controller.js]: Can you check if gulp build works on your code? You might need to disable/enable jshint camelcase around this line. See detail.ctrl how to do that..

File: app/templates/ui/app/create/create.html [hunk 6]
```diff
@@ -1,13 +1,47 @@
 <div class="create" ng-show="ctrl.currentUser && !ctrl.currentUser.disallowUpdates">
   <div class="row">
-    <div class="col-sm-2"></div>
-    <h2 class="col-sm-10">Create a Document</h2>
+    <div class="col-sm-2">
+    </div>
+    <h2 class="col-sm-10" ng-if="ctrl.mode == 'create'">Create a Document</h2>
+    <h2 class="col-sm-10" ng-if="ctrl.mode == 'edit'">Edit Document</h2>
```
Comment (grtjn) [inline app/templates/ui/app/create/create.html]: Consider showing `{{ctrl.uri}}` instead of `Document` in case of edit..

File: app/templates/ui/app/create/create.html [hunk 7]
```diff
@@ -1,13 +1,47 @@
...
+              <li ng-repeat="e in errors">{{ e.$name }} has an error: <strong>{{ key }}</strong>.</li>
+            </ul>
+          </li>
+        </ul>
+        -->
+     </div>
+    </div>
     <div class="form-group">
-      <label class="col-sm-2 control-label">Name</label>
+      <label class="col-sm-2 control-label">Name *</label>
```
Comment (grtjn) [inline app/templates/ui/app/create/create.html]: Use `required` that shows a red * with CSS instead of literally appending *..

File: app/templates/ui/app/detail/detail.controller.js [hunk 8]
```diff
@@ -40,7 +42,26 @@
...
+        // create a toast with settings:
+        ngToast.create({
+          className: 'warning',
+          content: 'Deleted ' + uri,
+          dismissOnTimeout: true,
+          timeout: 2000,
+          dismissButton: true,
+          onDismiss: function () {
+            //redirect to search page
+            $location.path('/search');
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.controller.js]: Use `$state.go` instead of `$location.path()`, see create.ctrl..

File: app/templates/ui/app/detail/detail.html [hunk 9]
```diff
@@ -1,27 +1,26 @@
...
-          <json-explorer ng-if="ctrl.type==='json'" data="ctrl.json"></json-explorer>
-          <pre ng-if="ctrl.type!=='json'" class="pre-xml">{{ctrl.xml}}</pre>
-
-          <a ng-if="ctrl.type==='binary'" target="_blank" ng-href="/v1/documents?uri={{ctrl.uri}}&transform=download" class="btn btn-default">Download</a>
-        </div>
-      </uib-tab>
-      <uib-tab ng-if="ctrl.type!=='json'">
+    <!-- note -->
+    <uib-tabset type="pills" active="ctrl.type">
+      <uib-tab index="'json'" ng-click="ctrl.type='json'" ng-if="ctrl.type!='binary'">
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.html]: show this if type === json

File: app/templates/ui/app/detail/detail.html [hunk 10]
```diff
@@ -1,27 +1,26 @@
...
         <uib-tab-heading>JSON</uib-tab-heading>
-        <div class="col-sm-9 bottom-buffer">
+        <div class="col-sm-9">
           <json-explorer data="ctrl.json"></json-explorer>
         </div>
       </uib-tab>
-      <uib-tab ng-if="ctrl.type==='json'">
+      <uib-tab index="'xml'" ng-click="ctrl.type='xml'" ng-if="ctrl.type!='binary'">
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.html]: show this if type === xml || type === text

File: app/templates/ui/app/detail/detail.html [hunk 11]
```diff
@@ -1,27 +1,26 @@
...
         <uib-tab-heading >XML</uib-tab-heading>
         <div class="col-sm-9">
           <pre class="pre-xml">{{ctrl.xml}}</pre>
         </div>
       </uib-tab>
+      <uib-tab index="'binary'" ng-click="ctrl.type='binary'" ng-if="ctrl.type=='binary'">
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.html]: Use === instead of ==

File: app/templates/ui/app/detail/detail.html [hunk 12]
```diff
@@ -34,3 +33,10 @@
     </div>
   </div>
 </div>
+<div class="row">
+  <div class="col-sm-12">
+    <button class="btn btn-primary" ui-sref="root.search">Cancel</button>
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.html]: Cancel should be btn-default..

File: app/templates/ui/app/detail/detail.html [hunk 13]
```diff
@@ -34,3 +33,10 @@
...
+    <button class="btn btn-primary" ui-sref="root.edit({uri: ctrl.uri})">Edit</button>
+    <button class="btn btn-primary" ng-click="ctrl.delete()">Delete</button>
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.html]: ctrl.delete() could show an 'are you sure' popup/popover if clicked. You can use angular-bootstrap for that..

File: app/themes/map/ui/app/app.js [hunk 14]
```diff
@@ -17,8 +17,10 @@
     'ui.bootstrap',
     'ui.router',
     'ui.tinymce',
+    'cb.x2js',
     'ngToast',
-    'uiGmapgoogle-maps'
+    'uiGmapgoogle-maps',
+    'ngMessages'
```
Comment (grtjn) [inline app/themes/map/ui/app/app.js]: Same here, should be in create.module.js only, which is not overridden by map theme, so I think you only need to change the default one, and just undo these changes..

File: app/themes/map/ui/app/create/create.html [hunk 15]
```diff
@@ -1,13 +1,37 @@
 <div class="create" ng-show="ctrl.currentUser && !ctrl.currentUser.disallowUpdates">
   <div class="row">
     <div class="col-sm-1"></div>
-    <h2 class="col-sm-11">Create a Document</h2>
+    <h2 class="col-sm-11" ng-if="ctrl.mode == 'create'">Create a Document</h2>
+    <h2 class="col-sm-11" ng-if="ctrl.mode == 'edit'">Edit Document</h2>
```
Comment (grtjn) [inline app/themes/map/ui/app/create/create.html]: again `{{ctrl.uri}}`..

File: app/themes/map/ui/app/create/create.html [hunk 16]
```diff
@@ -1,13 +1,37 @@
...
+            <ul>
+              <li ng-repeat="e in errors">{{ e.$name }} has an error: <strong>{{ key }}</strong>.</li>
+            </ul>
+          </li>
+        </ul>
+        -->
+      </div>
+    </div>
+    <div class="form-group">
+      <label class="control-label">Name *</label>
```
Comment (grtjn) [inline app/themes/map/ui/app/create/create.html]: again `required` class

File: app/themes/map/ui/app/detail/detail.controller.js [hunk 17]
```diff
@@ -58,13 +60,32 @@
...
+        // create a toast with settings:
+        ngToast.create({
+          className: 'warning',
+          content: 'Deleted ' + uri,
+          dismissOnTimeout: true,
+          timeout: 2000,
+          dismissButton: true,
+          onDismiss: function () {
+            //redirect to search page
+            $location.path('/search');
```
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.controller.js]: again `$state.go`..

File: app/themes/map/ui/app/detail/detail.html [hunk 18]
```diff
@@ -11,18 +12,30 @@
 
         </div>
       </uib-tab>
-      <uib-tab ng-if="ctrl.type!=='json'">
+      -->
+      <uib-tab ng-if="ctrl.type==='json'" ng-click="ctrl.type='json'"
+               ng-if="ctrl.type!='binary'">
```
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.html]: two ng-if's? Drop the second..

File: app/themes/map/ui/app/detail/detail.html [hunk 19]
```diff
@@ -11,18 +12,30 @@
...
         <uib-tab-heading>JSON</uib-tab-heading>
         <div class="col-sm-12">
           <json-explorer data="ctrl.json"></json-explorer>
         </div>
       </uib-tab>
-      <uib-tab ng-if="ctrl.type==='json'">
+      <uib-tab ng-if="ctrl.type==='xml'" ng-click="ctrl.type='xml'"
+               ng-if="ctrl.type!='binary'">
```
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.html]: dito..

File: app/themes/map/ui/app/detail/detail.html [hunk 20]
```diff
@@ -11,18 +12,30 @@
...
         <uib-tab-heading >XML</uib-tab-heading>
         <div class="col-sm-12">
           <pre class="pre-xml">{{ctrl.xml}}</pre>
         </div>
       </uib-tab>
+      <uib-tab ng-if="ctrl.type==='binary'" ng-click="ctrl.type='binary'"
+               ng-if="ctrl.type=='binary'">
```
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.html]: dito..

File: app/themes/map/ui/app/detail/detail.html (hunk 19 above)
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.html]: should also be shown for type === text

File: app/themes/map/ui/app/detail/detail.html [hunk 21]
```diff
@@ -35,3 +48,10 @@
     </div>
   </div>
 </div>
+<div class="row">
+  <div class="col-sm-12">
+    <button class="btn btn-primary" ui-sref="root.search">Cancel</button>
```
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.html]: btn-default..

File: app/themes/map/ui/app/detail/detail.html [hunk 22]
```diff
@@ -35,3 +48,10 @@
...
+    <button class="btn btn-primary" ui-sref="root.edit({uri: ctrl.uri})">Edit</button>
+    <button class="btn btn-primary" ng-click="ctrl.delete()">Delete</button>
```
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.html]: popover 'are you sure?'

File: slushfile.js [hunk 23]
```diff
@@ -273,6 +273,17 @@ function configRoxy() {
       '          <range-value-positions>false</range-value-positions>\n' +
       '        </geospatial-element-pair-index>\n');
 
+    // add range path index for the default content
+    foo = foo.replace(/^\s*<range-path-indexes>/m,
+        '      <range-path-indexes>\n' +
+        '        <range-path-index>\n' +
+        '          <scalar-type>string</scalar-type>\n' +
+        '          <collation>http://marklogic.com/collation/codepoint</collation>\n' +
+        '          <path-expression>//docFormat</path-expression>\n' +
```
Comment (grtjn) [inline slushfile.js]: you can skip the // in the path..

File: app/templates/ui/app/create/create.controller.js [hunk 24]
```diff
@@ -4,13 +4,47 @@
...
+      }
+
+      if (!ctrl.person.location) {
+        ctrl.person.location = {
+          latitude: 0,
+          longitude: 0
+        };
+      }
+
+      ctrl.uri = doc.config.params.uri;
```
Comment (grtjn) [inline app/templates/ui/app/create/create.controller.js]: doc.config.params.uri? Is that part of the content in the database like that? Isn't it easier to grab uri from `$stateParams`? (like in detail.ctrl)

File: app/templates/bower.json [hunk 25]
```diff
@@ -2,38 +2,43 @@
...
     "angular-ui-router": "~0.2.15",
     "angular-ui-tinymce": "~0.0.9",
-    "angular-animate": "~1.4.4",
-    "ngtoast": "^2.0.0",
     "angular-x2js": "https://github.com/janmichaelyu/angular-x2js.git",
     "bootstrap": "~3.3.5",
     "font-awesome": "~4.6.0",
+    "highcharts": "^4.2",
     "highlightjs":"~8.7.0",
+    "jquery": "~2.1.4",
```
Comment (grtjn) [inline app/templates/bower.json]: Thanks for alphabetizing this list. There is one exception though, jquery must be listed first (or at least before anything from Angular). That is to make sure Angular will use jquery, which is important for some of the reusable components we have put together..

File: app/themes/map/ui/app/detail/detail.html [hunk 26]
```diff
@@ -35,3 +36,10 @@
     </div>
   </div>
 </div>
+<div class="row">
+  <div class="col-sm-12">
+    <button class="btn btn-primary btn-default" ui-sref="root.search">Cancel</button>
```
Comment (grtjn) [inline app/themes/map/ui/app/detail/detail.html]: Can you drop btn-primary on this one? I don't think you can both have btn-primary and btn-default.. :)

File: app/templates/ui/app/detail/detail.html [hunk 27]
```diff
@@ -34,3 +33,10 @@
     </div>
   </div>
 </div>
+<div class="row">
+  <div class="col-sm-12">
+    <button class="btn btn-primary btn-default" ui-sref="root.search">Cancel</button>
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.html]: Drop btn-primary please..

File: app/templates/ui/app/create/create.html [hunk 28]
```diff
@@ -91,7 +133,7 @@ <h2 class="col-sm-10">Create a Document</h2>
     <div class="form-group">
       <label class="col-sm-2 control-label">Tags</label>
       <div class="col-sm-10">
-        <input type="text" ng-model="ctrl.newTag">
+        <input type="text" ng-model="ctrl.newTag"/>
```
Comment (grtjn) [inline app/templates/ui/app/create/create.html]: Can you remove the /? It is not necessary in html..

**Comment by grtjn** [comment]
Just a few more syntactic things, see inline comments. I'll try to run a test with this code in parallel. It looks really good at first glance..

File: app/templates/rest-api/config/options/all.xml [hunk 29]
```diff
@@ -92,11 +92,20 @@
...
   </constraint>
 
+  <constraint name="docFormat">
+    <range type="xs:string" facet="true" collation="http://marklogic.com/collation/codepoint">
+      <facet-option>limit=5</facet-option>
+      <facet-option>item-order</facet-option>
+      <facet-option>descending</facet-option>
+      <path-index>docFormat</path-index>
+    </range>
+  </constraint>
```
Comment (grtjn) [inline app/templates/rest-api/config/options/all.xml]: When you load the sample-data, this facet doesn't get populated. It would be nice if it would, but afraid that means touching all 3000 docs in that sample-data zip.. Hmm, didn't I write a nodejs script to fix lat/long in them?

File: app/templates/ui/app/create/create.html [hunk 30]
```diff
@@ -109,10 +151,29 @@ <h2 class="col-sm-10">Create a Document</h2>
...
+      <div class="col-md-10 col-md-offset-2 has-error has-feedback" >
+        <div ng-messages="createForm.docFormat.$error">
+          <p class="alert alert-danger" ng-message="required">Document Format is required.</p>
+        </div>
+      </div>
+    </div>
     <div class="row">
-      <a class="col-sm-offset-10 btn btn-default" ui-sref="root.landing">Cancel</a>
-      <button class="btn btn-primary" ng-click="ctrl.submit()">Create</button>
+      <a class="btn btn-default" ui-sref="root.landing">Cancel</a>
```
Comment (grtjn) [inline app/templates/ui/app/create/create.html]: Wondering, would it make sense to jump back to details if you are editing?

File: app/templates/ui/app/detail/detail.controller.js [hunk 31]
```diff
@@ -40,7 +42,41 @@
...
+            return $uibModalInstance.close();
+          };
+        }],
+        controllerAs: 'ctrl',
+        templateUrl: '/app/detail/confirm-modal.html',
+        backdrop: 'static',
+        keyboard: false
+      });
+    }
+
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.controller.js]: Ah right, I had something simpler in mind when I mentioned popover: https://angular-ui.github.io/bootstrap/#/popover. Showing a modal is perhaps a bit overkill. I used that myself as well before, but liked the popover better when I saw it being used instead recently on some ordinary webpage.

File: app/templates/ui/app/detail/detail.html [hunk 32]
```diff
@@ -34,3 +33,10 @@
     </div>
   </div>
 </div>
+<div class="row">
+  <div class="col-sm-12">
+    <button class="btn btn-default" ui-sref="root.search">Cancel</button>
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.html]: Using `Cancel` here looks a bit funny. How about `Back`, or `Back to (looking glass icon)`?

File: app/templates/ui/app/create/create.html [hunk 33]
```diff
@@ -62,9 +96,17 @@ <h2 class="col-sm-10">Create a Document</h2>
...
+      <label class="col-sm-2 control-label required">Email</label>
       <div class="col-sm-10">
-        <input type="email" class="form-control" ng-model="ctrl.person.email" placeholder="">
+        <input type="email" class="form-control" ng-model="ctrl.person.email" placeholder="" name="email" required>
+      </div>
+    </div>
+    <div class="row form-error" ng-if="(createForm.$submitted || createForm.email.$touched) && createForm.email.$error">
+      <div class="col-md-10 col-md-offset-2 has-error has-feedback" >
+        <div ng-messages="createForm.email.$error">
+          <p class="alert alert-danger" ng-message="required">Email is required.</p>
```
Comment (grtjn) [inline app/templates/ui/app/create/create.html]: Could you drop required on email, and make eyeColor required instead? You can leave the email type check in place, but it should only fire if something was entered..

File: app/templates/ui/app/create/create.html [hunk 34]
```diff
@@ -62,9 +96,17 @@ <h2 class="col-sm-10">Create a Document</h2>
...
+          <p class="alert alert-danger" ng-message="email">Invalid email format.</p>
+        </div>
       </div>
     </div>
```
Comment (grtjn) [inline app/templates/ui/app/create/create.html]: Come to think of it, it looks like there are a few other input fields with specific types. They are not validated, e.g. you can enter xxx for age..

Maybe also make 'Gender' required? That way things look a bit more realistic..

File: app/templates/ui/app/create/create.controller.spec.js [hunk 35]
```diff
@@ -27,7 +29,16 @@
 
     beforeEach(function () {
       // stub the current user
```
Comment (grtjn) [inline app/templates/ui/app/create/create.controller.spec.js]: Could you come up with a more appropriate comment?

File: app/templates/ui/app/create/create.controller.spec.js [hunk 36]
```diff
@@ -9,10 +9,12 @@
...
     beforeEach(function() {
       bard.appModule('app.create');
-      bard.inject('$controller', '$q', '$rootScope', 'MLRest', '$state', 'userService');
+      bard.inject('$controller', '$q', '$rootScope', 'MLRest', '$state', 'userService',
+        'x2js');
 
       bard.mockService(MLRest, {
-        createDocument: $q.when('/?uri=blah')
+        createDocument: $q.when('/?uri=blah'),
+        updateDocument: $q.when('/?uri=blah')
```
Comment (grtjn) [inline app/templates/ui/app/create/create.controller.spec.js]: Hmm, it is now only testing editing of a document I think. Maybe we need to create two copies of the controller, one for create and one for update, and duplicate the `should add tags` test case, one for create and one for update.. that should suffice..

File: app/templates/ui/app/detail/detail.module.js [hunk 37]
```diff
@@ -3,6 +3,9 @@
 
   angular.module('app.detail', [
     'app.similar',
-    'ui.router'
+    'ui.router',
+    'ml.common', // unit test fails without this
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.module.js]: better comment perhaps: // Using MLRest for delete

File: app/templates/ui/app/detail/detail.module.js [hunk 38]
```diff
@@ -3,6 +3,9 @@
...
+    'ngToast', // unit test fails without this
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.module.js]: // Showing a toast on delete

File: app/templates/ui/app/detail/detail.module.js [hunk 39]
```diff
@@ -3,6 +3,9 @@
...
+    'ui.bootstrap' // unit test fails without this
```
Comment (grtjn) [inline app/templates/ui/app/detail/detail.module.js]: if you change from modal to popover, you probably don't need this, but otherwise:
// Showing a modal for delete confirmation

File: app/themes/map/ui/app/create/create.html [hunk 40]
```diff
@@ -62,9 +87,15 @@ <h2 class="col-sm-11">Create a Document</h2>
...
-        <input type="email" class="form-control" ng-model="ctrl.person.email" placeholder="">
+      <label class="control-label required">Email</label>
+      <div>
+        <input type="email" class="form-control" ng-model="ctrl.person.email" placeholder=""
+               placeholder="" name="email" required>
+        <div ng-messages="createForm.email.$error"
+             ng-if="(createForm.$submitted || createForm.email.$touched) && createForm.email.$error">
+          <p class="alert alert-danger" ng-message="required">Email is required.</p>
+          <p class="alert alert-danger" ng-message="email">Invalid email format.</p>
+        </div>
```
Comment (grtjn) [inline app/themes/map/ui/app/create/create.html]: See comments on the default create.html file..

**Comment by grtjn** [comment]
@jenbreese There are still a few items being looked at, but I think it is stable enough for you to take a look at the UI. I'm particularly interested in what you think about button placement on Create/Edit and Details page, and about validation messages and such on Create/Edit.

Note: if you have linked slush-marklogic-node from your local fork, you could use part of the manual merge commands to pull them in easily:

```
git checkout -b divino-15-Edit-and-delete-options master
git pull https://github.com/divino/slush-marklogic-node.git 15-Edit-and-delete-options
```

Once done you switch back to one of your own branches, and you can delete this one..

File: app/templates/rest-api/config/options/all.xml (hunk 29 above)
Comment (grtjn) [inline app/templates/rest-api/config/options/all.xml]: sample-data.zip was updated for this..

File: app/templates/ui/app/create/create.controller.spec.js (hunk 36 above)
Comment (grtjn) [inline app/templates/ui/app/create/create.controller.spec.js]: changes were applied..

**Comment by grtjn** [comment]
Looking good! I'll run a test tomorrow morning..

**Comment by grtjn** [comment]
Nice find that mwl-confirm directive! I think the popover could use a slightly more informative message. How about "This will permanently delete {{ctrl.uri}}, are you sure?".

I'm also thinking we should make the buttons on the detail page appear more prominently. How about putting them above the similar panel (justified left and right?) on the regular detail page, and on the top in the map detail page.

Otherwise, it all works like a charm..

**Comment by grtjn** [comment]
Good enough for me. :)

@jenbreese Let me make it easier for you. I'm merging now, so you can just pull changes from master. If you have time, could you look at in particular the detail page of default and map theme?

**Comment by grtjn** [comment]
This includes a fix for #324 as well..
